jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/CloudFormationTemplate.kt

```diff
@@ -114,4 +118,25 @@ fun List<Parameter>.mergeRemoteParameters(remoteParameters: List<software.amazon
             mutableParameter.setDefaultValue(it.parameterValue())
         }
         mutableParameter
-    }.toList()
\ No newline at end of file
+    }.toList()
+
+/**
+ * Validate whether the Lambda function runtimes in the specified template are supported to build before deployment to AWS.
+ *
+ * @param virtualFile SAM template file
+ * @return null if they are supported, or an error message otherwise.
+ */
+fun Project.validateSamTemplateLambdaRuntimes(virtualFile: VirtualFile): String? {
+    val path = virtualFile.path
+    return CloudFormationTemplateIndex
+            .listFunctions(this, virtualFile)
+            .mapNotNull {
+                Runtime.fromValue(it.runtime()).validOrNull?.let { runtime ->
+                    runtime.runtimeGroup?.let { runtimeGroup ->
+                        if (runtimeGroup.supportSamBuild()) return@mapNotNull null
+                        else message("serverless.application.deploy.error.unsupported_runtime_group", runtime.toString(), path)
```

Comment:
Use intermediate variables for the sake of readability, this is super hard to follow